patches/server/1029-Reduce-work-done-in-CraftMapCanvas.drawImage-by-limi.patch

```diff
@@ -0,0 +1,92 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Barnaby <22575741+barnabwhy@users.noreply.github.com>
+Date: Sat, 29 Jun 2024 12:06:51 +0100
+Subject: [PATCH] Reduce work done in CraftMapCanvas.drawImage by limiting size
+ of image and using System.arraycopy instead of for loops and use bitwise
+ operations to do bounds checks.
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java b/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java
+index ff59f759669620795ef355c988b664bdcda39f52..54a3c1cea840f45385f1102988f8d2c8c415127b 100644
+--- a/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java
++++ b/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java
+@@ -59,8 +59,10 @@ public class CraftMapCanvas implements MapCanvas {
+ 
+     @Override
+     public void setPixel(int x, int y, byte color) {
+-        if (x < 0 || y < 0 || x >= 128 || y >= 128)
++        // Paper start - Optimise bounds check using bitwise operations
++        if ((x & -128) != 0 || (y & -128) != 0)
```

Comment:
Are you sure this is needed/did you benchmark this change? I can't imagine this will scrape off much even beyond a nanosecond, such simple integer comparisons are generally free, but this does make readability a lot worse

Reply:
I don't recall it made much difference to be honest, I can remove it if you feel that makes more sense.